juliet/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+#![doc = include_str!("../README.md")]
+
+//!
+//!
+//! ## General usage
+//!
+//! This crate is split into three layers, whose usage depends on an applications specific usecase.
+//! At the very core sits the [`protocol`] module, which is a side-effect free implementation of the
+//! protocol. The caller is responsible for all IO flowing in and out, but it instructed by the
+//! state machine what to do next.
+//!
+//! If there is no need to roll custom IO, the [`io`] layer provides a complete `tokio`-based
+//! solution that operates on [`tokio::io::AsyncRead`] and [`tokio::io::AsyncWrite`]. It handles
+//! multiplexing input, output and scheduling, as well as buffering messages using a wait and a
+//! ready queue.
+//!
+//! Most users of the library will likely use the highest level layer, [`rpc`] instead. It sits on
+//! top the raw [`io`] layer and wraps all the functionality in safe Rust types, making misuse of
+//! the underlying protocol hard, if not impossible.
+
```

Comment:
I added those except I used the `juliet` crate name for now, and had to remove the `unused_qualifications`, since it does not play nice with the `try_outcome!` macro: b259a2085